### PR TITLE
eventloop: zero priority linked list on init

### PR DESCRIPTION
**Pull request by matthewfala**
> Note to the reviewer: This change requires from now on that when creating an event in memory, `MK_EVENT_ZERO()` is called or else a segfault on `mk_event_add()` will occur.

## Problem
Currently if an event is added to the event loop twice, an error can occur where the ready list (priority list) is corrupted, as the link list in the event is zeroed.

## Solution Difficulties
Since mk_event may be uninitialized memory, we have no way of tracking whether the event has been added to the loop before

~~## Solution
Require that MK_EVENT_ZERO is always called before the mk_event_add() function is called. Then we know if we should remove records of the event from the loop before adding again.~~

~~## Honest Comments~~
~~Honestly, I'm not a huge fan of this solution, as it means that if we forget to zero an mk_event after creating it, the introduced code in `mk_event_add` will segfault, since it expects that the linked list is either 0 or valid, and will try to remove the event if non zero. ~~

~~However it was the most reasonable option we could think of. And we thought it would make sense to require some sort of initialization function for the mk_list's memory~~

~~## Other Options to Consider~~
~~1. Require that the same events are not added multiple times without being first deleted.
2. Add memory to the event loop context to track what events have been added. That way additional adds won't corrupt the event loop.
3. Introduce a new function to Fluent Bit called `mk_event_add_safe()` that checks the event loop's internal priority loops for references to the event and removes it before calling the normal mk_add. (this requires enumerating the list for each add)~~

## New Minimal Solution
In order to reduce the potential for this change to break things, and following suggestions of Leonardo, we are changing all event memory initialization to calloc (mk_mem_alloc_z). We will assume that on creation, the linked list _priority_head in the event will be set to 0. This will be maintained by mk_event_add and mk_event_del. MK_EVENT_ZERO, MK_EVENT_NEW, MK_EVENT_INIT are not changed.

## To Keep in Mind
From now on, wherever memory for an event is allocated, that event's memory should be zero.
This [Fluent Bit PR](https://github.com/fluent/fluent-bit/pull/7088) ensures that on memory allocation, event memory is initialized to zero.

## MK Event Proper Event Initialization Notes
Allocation of events before MK_EVENT_INIT, MK_EVENT_ZERO, MK_EVENT_NEW (ie, outside of mk_core/)

MK_EVENT_INIT - None
MK_EVENT_ZERO - None
MK_EVENT_NEW

    mk_fifo.c - fw->event
        changed fw to be allocated with zero
    mk_http_thread.c - channel->event (event is a pointer)
        Finding where this event is allocated is tricky. The event pointer is set in mk_sched_add_connection from conn which is allocated with mk_mem_alloc_z on line 200 of mk_scheduler.c, see: https://github.com/fala-aws/monkey/blob/67d9fd61e1ff294f5c08f69444d0861639ea4286/mk_server/mk_scheduler.c#L200
    mk_net.c
        The first 2 instances are from conn which is an mk_net_connection holding an event
            change conn to be allocated with mk_mem_alloc_z
        The second 2 instances are from channel->event which is a pointer. This is same as mk_http_thread. The event is initialized via mk_mem_alloc_z()

Summary of changes:

    mk_mem_alloc -> mk_mem_alloc_z for fw, containing an event
    mk_mem_alloc -> mk_mem_alloc_z for mk_net_connection, containing an event

### CODE REVIEW

**Comment by matthewfala** [comment]
Todo for Matt: I need to add a bunch of MK_EVENT_ZEROS to the code.

@leonardo-albertovich  would it be possible for you to take a look at this? Are you okay requiring that after creating the event in memory, the MK_EVENT_ZERO() is called or else a segfault on mk_event_add() will occur?

**Review by leonardo-albertovich — CHANGES_REQUESTED** [review]
Once you make the requested change pleaase let us know, this is an important fix that warrants a monkey update in 1.9. 2.0 and master in the current release cycle.

**Comment by leonardo-albertovich** [comment]
> 

I looked at the changes you made and they add up to me. 

About event creation, I think `MK_EVENT_NEW` is the macro used for that and frankly I think it should just be an alias to `MK_EVENT_INIT` (maybe zeroing out the memory before just in case someone allocated the event with something that doesn't initialize memory or if it's located in the stack).

I suppose the reason why you mention `_mk_event_add` causing a crash would be having garbage pointers in those the list linkage structure and if that's the case then I'm entirely sure it should crash and do it in a very obvious and loud way so we know some client code is being sloppy which we should catch in CI or integration tests before it's released.

My overall posture when it comes to this kind of situation is that the code should never hide the bugs and yes, I know, I sound like captain obvious but sometimes we try to be helpful or smart by handling these cases and in the end it causes us more headaches than a crash or error message would.

File: include/monkey/mk_core/mk_event.h [hunk 1]
```diff
@@ -122,6 +122,8 @@ static inline void MK_EVENT_INIT(struct mk_event *ev, int fd, void *data,
 static inline void MK_EVENT_ZERO(struct mk_event *e)
 {
     MK_EVENT_INIT(e, -1, NULL, NULL);
+    e->_priority_head.next = NULL;
```
[the remark this answers is not in the document]
Comment (matthewfala) [reply include/monkey/mk_core/mk_event.h:125]: This sounds good. Thank you.

File: include/monkey/mk_core/mk_event.h [hunk 2]
```diff
@@ -117,17 +117,20 @@ static inline void MK_EVENT_INIT(struct mk_event *ev, int fd, void *data,
     ev->status  = MK_EVENT_NONE;
     ev->data    = data;
     ev->handler = callback;
+    mk_list_entry_init(&e->_priority_head);
```
Comment (matthewfala) [inline include/monkey/mk_core/mk_event.h:120]: should be ev

File: include/monkey/mk_core/mk_event.h [hunk 3]
```diff
@@ -117,17 +117,20 @@ static inline void MK_EVENT_INIT(struct mk_event *ev, int fd, void *data,
...
 }
 
 static inline void MK_EVENT_ZERO(struct mk_event *e)
 {
     MK_EVENT_INIT(e, -1, NULL, NULL);
+    mk_list_entry_init(&e->_priority_head);
```
Comment (matthewfala) [inline include/monkey/mk_core/mk_event.h:126]: not needed

**Comment by matthewfala** [comment]
Notes:

### Allocation of events before MK_EVENT_INIT, MK_EVENT_ZERO, MK_EVENT_NEW (ie, outside of mk_core/)
MK_EVENT_INIT - None
MK_EVENT_ZERO - None
MK_EVENT_NEW
- mk_fifo.c - fw->event
   - changed fw to be allocated with zero
- mk_http_thread.c - channel->event (event is a pointer)
   - Finding where this event is allocated is tricky. The event pointer is set in `mk_sched_add_connection` from conn which is allocated with mk_mem_alloc_z on line 200 of mk_scheduler.c, see: https://github.com/fala-aws/monkey/blob/67d9fd61e1ff294f5c08f69444d0861639ea4286/mk_server/mk_scheduler.c#L200
- mk_net.c
   - The first 2 instances are from conn which is an mk_net_connection holding an event
      - change conn to be allocated with mk_mem_alloc_z
   - The second 2 instances are from channel->event which is a pointer. This is same as mk_http_thread. The event is initialized via mk_mem_alloc_z()

Summary of changes:
- mk_mem_alloc -> mk_mem_alloc_z for fw, containing an event
- mk_mem_alloc -> mk_mem_alloc_z for mk_net_connection, containing an event

**Comment by matthewfala** [comment]
@leonardo-albertovich  Thanks for your patience. Would you take another look?

**Comment by matthewfala** [comment]
Thank you!

**Comment by edsiper** [comment]
@leonardo-albertovich is this ok to merge it to update Fluent Bit ?

**Comment by leonardo-albertovich** [comment]
Yes!

**Pull request merged**